src/contracts/caching_log_query.rs

```diff
@@ -0,0 +1,214 @@
+use crate::{
+    database::{Database, DatabaseError},
+    ethereum::ProviderStack,
+};
+use async_stream::try_stream;
+use ethers::{
+    providers::{Middleware, ProviderError},
+    types::{Filter, Log, U64},
+};
+use futures::Stream;
+use serde_json::value::RawValue;
+use std::{
+    cmp::{max, min},
+    sync::Arc,
+};
+use thiserror::Error;
+
+pub struct CachingLogQuery {
+    provider:           Arc<ProviderStack>,
+    filter:             Filter,
+    page_size:          u64,
+    cache_blocks_delay: u64,
+    database:           Option<Arc<Database>>,
+}
+
+#[derive(Error, Debug)]
+pub enum Error<ProviderError> {
+    #[error(transparent)]
+    LoadLastBlock(ProviderError),
+    #[error(transparent)]
+    LoadLogs(ProviderError),
```

Comment:
We have two different wrappers for `ProviderError`, but both are `#[transparent]`. This has the consequence that there is no way to distinguish the two cases based on the error message. It would be better to distinguish:

```
    #[error("Error while loading last block: {0}")]
    LoadLastBlock(#[source] ProviderError),
    #[error("Error while loading last block: {0}")]
    LoadLogs(#[source] ProviderError),
```

If we don't use the `#[from]` attribute, we should still set the `#[source]` attribute so the error-chain is preserved.